### PR TITLE
UW 368

### PR DESCRIPTION
**Pull request by MYWeidner**
**Description**
Stage cycle-dependent files in Forecast methods. Edit `stage_static_files` to make it `stage_files`. Arguments to method include `run_directory` (str), `files_to_stage` (dict – keys are file names relative run_directory, values are absolute paths (str) to source files, symbolic link = False). Copy or link the files. Default to copy.

Existing test tweaked to reflect name change. Added parametrization to tests (options: static and cycle dependent). New test to test symbolic link functionality.

**Type of change**

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**
- [x] pytests in GitHub actions

**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation.  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

**Review by WeirAE — COMMENTED** [review]
Distracted mid-conference look seems ok - nicely done, will recheck next break and should move to approve!

File: src/uwtools/tests/drivers/test_forecast.py [hunk 1]
```diff
@@ -174,26 +175,40 @@ def test_forecast_run_cmd():
     Tests that the command to be used to run the forecast executable was built successfully.
     """
     hera_expected = "srun --export=ALL test_exec.py"
-    assert hera_expected == FV3Forecast().run_cmd("--export=ALL", run_cmd="srun", exec_name="test_exec.py")
+    assert hera_expected == FV3Forecast().run_cmd(
```
Comment (WeirAE) [inline src/uwtools/tests/drivers/test_forecast.py:178]: Looks like the linter may have gotten carried away here, I've had it do this to me. Double check your environment again and it should set it back

**Review by WeirAE — CHANGES_REQUESTED** [review]
Other than reverting the extra formatting, I think this fulfills the requirements nicely!

File: src/uwtools/tests/drivers/test_forecast.py (hunk 1 above)
Comment (MYWeidner) [reply src/uwtools/tests/drivers/test_forecast.py]: Fixed the formatting issue - copy/paste the code from original.

**Review by WeirAE — APPROVED** [review]
Thanks for the quick fix, LGTM! Just wait for a second look and you're good to go!

File: src/uwtools/tests/drivers/test_forecast.py [hunk 2]
```diff
@@ -207,7 +209,35 @@ def test_FV3Forecast_stage_static_files(tmp_path):
...
     forecast_obj.create_directory_structure(run_directory)
-    forecast_obj.stage_static_files(run_directory, files_to_stage)
+    forecast_obj.stage_files(run_directory, files_to_stage)
     # Test that all of the destination files now exist:
     for dst_fn in files_to_stage.keys():
         assert (run_directory / dst_fn).is_file()
+
+
+@pytest.mark.parametrize("section", ["static", "cycledep"])
+def test_FV3Forecast_stage_files_link(tmp_path, section):
```
Comment (maddenp-noaa) [inline src/uwtools/tests/drivers/test_forecast.py:219]: I think we might be able to combine this test with the previous one by stacking parameterizations, e.g.
```
@pytest.mark.parametrize("section", ["static", "cycledep"])
@pytest.mark.parametrize("link_files", [True, False])
def test_FV3Forecast_stage_files(link_files, section, tmp_path):
...
```
(I'm not sure if the argument order matters -- I suspect not), then doing the call as
```
forecast_obj.stage_files(run_directory, files_to_stage, link_files)
```
then using the `link_files` parameter to decide whether to call `.is_file()` or `.is_symlink()`. That would only be a couple more lines of code in one method to potentially remove more duplicate code.

**Review by maddenp-noaa — APPROVED** [review]
Looks good! I made one suggestion about combining two unit tests; maybe see if it makes sense to you and seems worth trying.

You probably already have, but a final `make format && make test` in your development shell (and committing any changes made by `make format`) wouldn't hurt. I only mention it because it's a new thing, not because there appear to be any issues.

File: src/uwtools/tests/drivers/test_forecast.py (hunk 2 above)
Comment (MYWeidner) [reply src/uwtools/tests/drivers/test_forecast.py]: Stacked parameters and now we have one test_stage_files() that tests both copy and linking functionality. Ran make format && make test. Formatter updated the format for test_forecast_run_cmd(). As I understand it, we should allow the formatter to make these changes.

Comment (maddenp-noaa) [reply src/uwtools/tests/drivers/test_forecast.py]: Very nice.

Yes, it's appropriate at this point to commit any changes made by `make format`. In fact, we've got a ticket in the backlog to add a CI step that tests that `make format` run in CI makes _no_ changes, meaning that we'll expect developers to have run `make format` and committed the changes themselves.

**Review by maddenp-noaa — APPROVED** [review]

File: src/uwtools/tests/drivers/test_forecast.py (hunk 2 above)
Comment (WeirAE) [reply src/uwtools/tests/drivers/test_forecast.py]: I think the run_cmd reformatting may be an issue with your environment not reading the correct settings for the formatter.

**Review by WeirAE — APPROVED** [review]
Yep, makes sense, LGTM!

**Pull request merged**